File: August-LeetCoding-Challenge/09-Rotting-Oranges/Rotting-Oranges.cpp

const int dx[4] = { -1, 0, 1, 0 };
const int dy[4] = { 0, -1, 0, 1 };
typedef pair<int,int> ip;
class Solution {
public:
    int orangesRotting(vector<vector<int>>& grid) {
        const int m=grid.size(), n=grid[0].size();
        unordered_set<int> freshes;
        queue<ip> rottens;
        int r=0;
        for(int i=0; i<m; ++i){
            for(int j=0; j<n; ++j){
                if(grid[i][j]){
                    if(grid[i][j]==1) freshes.insert(n*i+j);
                    else rottens.emplace(i,j);
                }
            }
        }
        while(!rottens.empty()){
            for(int i=0, qsz=rottens.size(); i<qsz; ++i){
                ip f=rottens.front(); rottens.pop(); auto [x,y] = f;
                for(int j=0; j<4; ++j){
                    int tox=dx[j]+x, toy=dy[j]+y, ind = n*tox+toy;
                    if(tox>=0 && toy>=0 && tox<m && toy<n && freshes.count(ind)) 
                        freshes.erase(ind), rottens.emplace(tox,toy);
                }
                
            }
            if(!rottens.empty())
                ++r;           
        }
        if(!freshes.empty()) return -1;
        else return r;
    }
};

File: August-LeetCoding-Challenge/09-Rotting-Oranges/Rotting-Oranges.py
class Solution:
    
    def __init__(self):
        self.res = 0
        
    def orangesRotting(self, grid: List[List[int]]) -> int:
        q = []
        for i in range(len(grid)):
            for j in range(len(grid[0])):
                if grid[i][j]==2:
                    q.append([i,j])
        self.bfs(grid,q)                    
        for i in range(len(grid)):
            for j in range(len(grid[0])):
                if grid[i][j]==1:
                    return -1
        return self.res
                
    def bfs(self,grid,q):
        while q!=[]:
            l = []
            while q!=[]:
                l.append(q.pop(0))
            for tmp in l:
                a,b = tmp[0],tmp[1]
                length = len(q)
                if a+1<len(grid):
                    if grid[a+1][b]==1:
                        grid[a+1][b]=2
                        q.append([a+1,b])
                if a-1>=0:
                    if grid[a-1][b]==1:
                        grid[a-1][b]=2
                        q.append([a-1,b])
                if b+1<len(grid[0]):
                    if grid[a][b+1]==1:
                        grid[a][b+1]=2
                        q.append([a,b+1])
                if b-1>=0:
                    if grid[a][b-1]==1:
                        grid[a][b-1]=2
                        q.append([a,b-1])
            if len(q) > 0:
                self.res+=1
